ocaml/parsing/builtin_attributes.mli

```diff
@@ -199,3 +206,52 @@ val jkind : Parsetree.attributes -> jkind_attribute Location.loc option
     There should be at most one "error_message" attribute, additional ones are sliently
     ignored. **)
 val error_message_attr : Parsetree.attributes -> string option
+
+(** [get_payload] is a helper function for working with attribute payloads.
+    Given a payload that consists of a structure containing a single
+    [Pstr_eval], it will call the given function on the inner expression.
+    Otherwise, it returns [Error ()]. *)
+val get_payload :
+  (Parsetree.expression -> ('a, unit) Result.t) -> Parsetree.payload ->
+  ('a, unit) Result.t
+
+(** [get_optional_payload] is like [get_payload], except that in the case that
+    the payload consists of an empty structure, it returns [Ok None] rather than
+    [Error ()] . *)
+val get_optional_payload :
+  (Parsetree.expression -> ('a, unit) Result.t) -> Parsetree.payload ->
+  ('a option, unit) Result.t
+
+(** [get_id_from_exp] extracts the string from an identifier expression.
+    Suitable for use with [get_payload]. *)
+val get_id_from_exp : Parsetree.expression -> (string, unit) Result.t
+
+(* Support for property attributes like zero_alloc *)
+type property =
+  | Zero_alloc
+
+type check_attribute =
+  | Default_check
+  | Ignore_assert_all of property
+  | Check of { property: property;
+               strict: bool;
+               (* [strict=true] property holds on all paths.
+                  [strict=false] if the function returns normally,
+                  then the property holds (but property violations on
+                  exceptional returns or divering loops are ignored).
+                  This definition may not be applicable to new properties. *)
+               opt: bool;
+               loc: Location.t;
+             }
+  | Assume of { property: property;
+                strict: bool;
+                loc: Location.t;
+                never_returns_normally: bool;
+              }
+
+val is_check_enabled : opt:bool -> property -> bool
```

Comment:
"check" and "property" are two names for one thing, used interchangeably in several of these bindings. I find both names to be too vague.

I wonder if this signature's names are just too general. There's a variant, `property`, with a single constructor `Zero_alloc`. Could we get rid of that variant and change these names to say `_zero_alloc`, or alternatively just make a `Zero_alloc` submodule?

It's not clear to me whether we will ever want to extend `property`. Maybe we do and my inclination is just wrong here.

Reply:
(And I recognize you're just moving bindings around in this commit. But rehoming them is a fine enough excuse for me to bikeshed their names.)

Reply:
I agree the attempt at generality here leads to code that is more confusing than it could be, and am open to eliminating the premise that there is more than one `property`, at least until there actually is.  But I presume @gretay-js made it this way for a reason, perhaps she can weigh in.

Reply:
Yeah, we had other properties at the beginning (like check that a functio nhas no caml_apply).  None of them was so I'm happy to rename everything to `Zero_alloc_checker` etc and eliminate unnecessary but it should be done in a separate PR across the entire code not just frontent. I'm happy to make such a  PR.

> "check" and "property" are two names for one thing, used interchangeably in several of these bindings. I find both names to be too vague.

The intented meaning of "property"  and "check" were different, but I agree that "check" is vague and would love to hear your suggestion of a better name!

Reply:
Luckily the name `zero_alloc_checker` as you suggest sounds great to me :) Thanks for the context, I agree the original decision made sense but revisiting it now sounds good to me.

Reply:
Here is a draft PR for the proposed refactoring https://github.com/ocaml-flambda/flambda-backend/pull/2416. It will conflict with this one, so we should probably wait.